### DISCUSSION TITLE
Change license to GPL3 or reduce features?

### DISCUSSION THREAD

**Discussion by yajo** [copier-org/copier#1416] [Polls]
Hello dear community. As you can see in https://github.com/copier-org/copier/issues/1398 and https://github.com/orgs/copier-org/discussions/1397, we have a legal issue right now.

The problem is that 2 of our dependencies are GPL3-licensed. This means in practice that Copier is GPL3 as of today, by license infection. However, Copier always wanted to be MIT.

The dependencies and their use:
- jinja2-ansible-filters. These are basic filters shipped with copier. One of the most used is `to_nice_yaml`, but there are plenty more. We can remove this dependency and just provide our own MIT implementation of that filter.
- pyyaml-include. It is used to be able to include external `copier.yaml`s into a top-level one. Allows for reusability. We can remove this dependency and tell users to use meta-templates if they need reusability, or implement a different MIT extensibility system.

I haven't investigated if there are already alternate implementations on pypi. If there were any, the choice would be clear: change our dependency.

Otherwise, we only have 2 options. Depending on the use cases for Copier across our user base, we'll have to go one or the other route.

Thanks for answering! Please drop in your thoughts.

**Comment by sisp** [comment]
I'd prefer keeping Copier MIT-licensed because it enables a broader range of use case.

`pyyaml-include` implements readers for several non-YAML formats, which we don't need. I believe reimplementing its basic functionality, that we need for Copier, is doable.

`jinja2-ansible-filters` may involve more work to reimplement, but many of the filters look like tiny Jinja filter wrappers around basic Python stdlib functions, so it might not be too bad. I rather wonder about the legal situation when reimplementing such basic filters under a different license for which there is essentially only one way to implement them. The main difference might be the tests implementation then. An argument in favor of replacing this dependency might be little activity – there might not be much need for maintenance, but it looks a bit too inactive for my taste.

**Reply by yajo** [reply]
pyyaml-include could be replaced by a custom key in config: `_extends: []`. We shouldn't need to dive into pyyaml just for this. pyyaml-include was just the easiest way to do that, but not the only one.

Regarding filters, the only one we *really* need is `to_nice_yaml`, which is scattered all accross the `{{ _copier_answers|to_nice_yaml -}}`. However, jinja provides `tojson`, and YAML is a superset of JSON, so we could ask our users to replace that. Fixing https://github.com/copier-org/copier/issues/983 could help paving the way for that change.

**Reply by AdrianFreundQC** [reply]
I looked at the list of filters in `jinja2-ansible-filters` and most look quite simple to implement.

I guess the safest way would be to have one person implement tests and verify them against the original implementation and have another person that has never seen the original source implement a replacement.

**Reply by yajo** [reply]
Tests are there.

IMHO the best solution would be that there exists a drop-in replacement library for each of these, so we just change the dependencies and move on (changing some bits here and there).

**Comment by sisp** [comment]
In case we decide to keep Copier MIT-licensed, it may be a good idea to continuously check the licenses of our dependencies. A tool like [`pip-licenses`](https://github.com/raimon49/pip-licenses) might help, e.g. with its [`--allow-only`](https://github.com/raimon49/pip-licenses#option-allow-only) flag and a list of MIT-compatible licenses.
